test/regress.c

```diff
@@ -1260,6 +1260,9 @@ test_signal_pipeloss(void)
  * make two bases to catch signals, use both of them.  this only works
  * for event mechanisms that use our signal pipe trick.	 kqueue handles
  * signals internally, and all interested kqueues get all the signals.
+ * This is not expected to work with signalfd - having more than one
+ * descriptor in attempt to accept the same signal (or intersecting sets
+ * of signals) is not the thing signalfd() was designed for.
```

Comment:
The problem is not `signalfd`, but `SIG_UNBLOCK`

Reply:
Why `SIG_UNBLOCK`? The limitation is clearly stated in `signalfd()` manual page as:
```
If a signal appears in the mask of more than one of the file descriptors, then occurrences of that signal can be read (once) from any one of the file descriptors.
```
IOW if you `poll()` more than one descriptor in attempt to accept the same signal, all of them will be marked active (with `POLLIN` in `revents` after `poll()` return) on signal arrival, but you can read `struct signalfd_siginfo` from the only one of them, and attempt to read another active descriptor(s) always returns -1.

Reply:
Here I'm talking about the test, look:

```sh
$ strace -e/signal,/proc regress --no-fork signal/signal_switchbase
signal/signal_switchbase: rt_sigprocmask(SIG_BLOCK, [USR1], NULL, 8) = 0
signalfd4(-1, [USR1], 8, SFD_CLOEXEC|SFD_NONBLOCK) = 11
rt_sigprocmask(SIG_BLOCK, [USR1], NULL, 8) = 0
signalfd4(-1, [USR1], 8, SFD_CLOEXEC|SFD_NONBLOCK) = 12
rt_sigprocmask(SIG_UNBLOCK, [USR1], NULL, 8) = 0
--- SIGUSR1 {si_signo=SIGUSR1, si_code=SI_USER, si_pid=32667, si_uid=1000} ---
+++ killed by SIGUSR1 +++
User defined signal 1
```

Reply:
So because there is no way to tell that this signal had been used in another base, this does not work.
But this is OK.